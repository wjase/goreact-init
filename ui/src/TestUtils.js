import { act } from 'react-dom/test-utils';

export const waitForComponentToPaint = async (wrapper) => {
    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
        wrapper.update();
    });
};
