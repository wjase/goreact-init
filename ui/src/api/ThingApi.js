import axios from 'axios';

const getThings = () => {
    var promise = new Promise(function (resolve, reject) {
        // do a thing, possibly async, thenâ€¦
        axios
            .get('/api/v1/things')
            .then(response => {
                resolve(response.data)
            })
            .catch(() => {
                reject(Error('ERROR'))
            });
    });
    return promise;
}

export const ThingApi = {
    getThings:getThings
}

