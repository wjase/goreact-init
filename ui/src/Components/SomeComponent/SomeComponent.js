import React from 'react';
import { thingApi } from '../../api/ThingApi.js'

export const dataReducer = (state, action) => {
  if (action.type === 'ERROR_UPDATED') {
    return { ...state, things: [], error: true };
  }

  if (action.type === 'THINGS_UPDATED') {
    return { ...state, things: action.things, error: false };
  }

  throw new Error();
};

export const useThings = (initialData)=>{
  const [data, dispatch] = React.useReducer(dataReducer, initialData);
   
  React.useEffect( () => {
    thingApi.getThings()
      .then(things => {
        dispatch({ type: 'THINGS_UPDATED', things });
      }).catch((err) => {
        dispatch({ type: 'ERROR_UPDATED', err });
      });
  }, []);

  return data
}

export const SomeComponent = ({initialData}) => {

  let data = useThings(initialData)

  return (
    <div>
      <h2>My Async Data</h2>

      {data.error && <div className="error">Error</div>}

      <ul>
        {data.things.map(item => (
          <li key={item.objectID}>{item.title}</li>
        ))}
      </ul>
    </div>
  );
};


export default SomeComponent;