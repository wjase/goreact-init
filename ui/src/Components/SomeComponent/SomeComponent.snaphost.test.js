import React from 'react';
import { mount } from 'enzyme';
import { thingApi} from '../../api/ThingApi.js'

import SomeComponent from './SomeComponent';
import { act } from '@testing-library/react';
import {waitForComponentToPaint} from '../../TestUtils.js'

const initialData = {
    things: [],
    error: null,
};

describe('SomeComponent', () => {
    it('should render correctly ', async () => {

        let testArticles = new Promise(function (resolve, reject) {
            resolve([
                { "objectID": 1, "title": "title1" },
                { "objectID": 2, "title": "title2" },
                { "objectID": 3, "title": "title3" }
            ]);
        });
        const getArticlesSpy = jest.spyOn(thingApi, 'getThings');
        getArticlesSpy.mockReturnValue(testArticles);

        const component = mount(<SomeComponent initialData={initialData} />);
        await waitForComponentToPaint(component);
        expect(component).toMatchSnapshot();

        //teardown
        getArticlesSpy.mockClear();

    });
});
