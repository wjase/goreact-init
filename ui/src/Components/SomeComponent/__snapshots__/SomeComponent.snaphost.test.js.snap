// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SomeComponent should render correctly  1`] = `
<SomeComponent
  initialData={
    Object {
      "articles": Array [],
      "error": null,
    }
  }
>
  <div>
    <h2>
      My Async Data
    </h2>
    <ul>
      <li
        key="1"
      >
        title1
      </li>
      <li
        key="2"
      >
        title2
      </li>
      <li
        key="3"
      >
        title3
      </li>
    </ul>
  </div>
</SomeComponent>
`;
